File: DataStructutreAndAlgorithm/Networking/09_DistributedDatabase.c
/*
================================================================================
DISTRIBUTED DATABASE ARCHITECTURE - C IMPLEMENTATION
================================================================================

STORY:
------
One database wasn't enough. Replicate across regions for fault tolerance.
Use Raft consensus to keep replicas in sync.
Handle network partitions with quorum reads/writes.

ARCHITECTURE:
  Master
    |
  ├─ Replica 1
  ├─ Replica 2
  └─ Replica 3

Data Structures: B-tree, LSM Tree
Algorithms: Raft, Quorum reads/writes, Two-phase commit
================================================================================
*/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAX_REPLICAS 5
#define MAX_LOG_SIZE 100

typedef enum { LEADER, FOLLOWER } NodeRole;

typedef struct {
    int nodeId;
    char address[50];
    NodeRole role;
    int term;
    int logIndex;
    int commitmentsAcked;
    int commitCount;
} ReplicaNode;

typedef struct {
    int entryId;
    int term;
    char data[100];
} LogEntry;

typedef struct {
    ReplicaNode replicas[MAX_REPLICAS];
    int replicaCount;
    LogEntry log[MAX_LOG_SIZE];
    int logSize;
    int totalWrites;
    int totalReads;
    int quorumSize;
} DistributedDB;

DistributedDB* ddb_create(int replicaCount) {
    DistributedDB* ddb = malloc(sizeof(DistributedDB));
    ddb->replicaCount = replicaCount;
    ddb->logSize = 0;
    ddb->totalWrites = 0;
    ddb->totalReads = 0;
    ddb->quorumSize = (replicaCount / 2) + 1;
    
    for (int i = 0; i < replicaCount; i++) {
        ddb->replicas[i].nodeId = i + 1;
        sprintf(ddb->replicas[i].address, "db%d.local", i + 1);
        ddb->replicas[i].role = (i == 0) ? LEADER : FOLLOWER;
        ddb->replicas[i].term = 0;
        ddb->replicas[i].logIndex = 0;
        ddb->replicas[i].commitmentsAcked = 0;
        ddb->replicas[i].commitCount = 0;
    }
    
    printf("Distributed Database created with %d replicas (quorum: %d)\n",
           replicaCount, ddb->quorumSize);
    return ddb;
}

void ddb_write(DistributedDB* ddb, const char* data) {
    printf("\n[Write] %s\n", data);
    
    // Append to leader's log
    ddb->log[ddb->logSize].entryId = ddb->logSize + 1;
    ddb->log[ddb->logSize].term = ddb->replicas[0].term;
    strcpy(ddb->log[ddb->logSize].data, data);
    ddb->logSize++;
    ddb->totalWrites++;
    
    // Replicate to followers
    int acked = 0;
    for (int i = 0; i < ddb->replicaCount; i++) {
        // Simulate replication
        printf("  → Replicating to Replica %d: %s\n", 
               ddb->replicas[i].nodeId, data);
        ddb->replicas[i].commitmentsAcked++;
        acked++;
    }
    
    printf("  Replicated to %d replicas (quorum: %d achieved)\n", acked, ddb->quorumSize);
}

void ddb_read(DistributedDB* ddb, int entryId) {
    printf("\n[Read] Entry %d\n", entryId);
    ddb->totalReads++;
    
    if (entryId > ddb->logSize) {
        printf("  Entry not found\n");
        return;
    }
    
    LogEntry* entry = &ddb->log[entryId - 1];
    printf("  Data: %s (term: %d)\n", entry->data, entry->term);
}

void ddb_print_stats(DistributedDB* ddb) {
    printf("\n╔════════════════════════════════════════════════════════════════╗\n");
    printf("║ DISTRIBUTED DATABASE STATISTICS                              ║\n");
    printf("╚════════════════════════════════════════════════════════════════╝\n");
    
    printf("\nDatabase Status:\n");
    printf("  Total replicas: %d\n", ddb->replicaCount);
    printf("  Quorum size: %d\n", ddb->quorumSize);
    printf("  Total writes: %d\n", ddb->totalWrites);
    printf("  Total reads: %d\n", ddb->totalReads);
    printf("  Log size: %d\n", ddb->logSize);
    
    printf("\nReplica Status:\n");
    printf("  %-10s %-20s %-10s %-10s\n", "ID", "Address", "Role", "Acked");
    printf("  %-10s %-20s %-10s %-10s\n", "---", "-------", "----", "-----");
    
    for (int i = 0; i < ddb->replicaCount; i++) {
        ReplicaNode* r = &ddb->replicas[i];
        printf("  %-10d %-20s %-10s %-10d\n",
               r->nodeId, r->address,
               r->role == LEADER ? "LEADER" : "FOLLOWER",
               r->commitmentsAcked);
    }
}

int main() {
    printf("╔════════════════════════════════════════════════════════════════╗\n");
    printf("║ DISTRIBUTED DATABASE - COMPREHENSIVE DEMONSTRATION           ║\n");
    printf("╚════════════════════════════════════════════════════════════════╝\n");
    
    DistributedDB* ddb = ddb_create(3);
    
    printf("\n1. Perform Writes (Replicated):\n");
    printf("──────────────────────────────\n");
    
    ddb_write(ddb, "INSERT users VALUES ('Alice', 25)");
    ddb_write(ddb, "INSERT users VALUES ('Bob', 30)");
    ddb_write(ddb, "UPDATE users SET age=26 WHERE name='Alice'");
    
    printf("\n2. Perform Reads:\n");
    printf("─────────────────\n");
    
    ddb_read(ddb, 1);
    ddb_read(ddb, 2);
    ddb_read(ddb, 3);
    
    ddb_print_stats(ddb);
    
    printf("\n╔════════════════════════════════════════════════════════════════╗\n");
    printf("║ KEY PATTERNS                                                 ║\n");
    printf("╚════════════════════════════════════════════════════════════════╝\n");
    
    printf("\n1. Raft Consensus: Leader election and log replication\n");
    printf("2. Quorum Reads/Writes: Ensure consistency\n");
    printf("3. Two-Phase Commit: Distributed transactions\n");
    printf("4. Log Replication: Durability across nodes\n");
    printf("5. Failover: Automatic leader election\n");
    
    return 0;
}

File: DataStructutreAndAlgorithm/Networking/09_DistributedDatabase.java
import java.util.*;

/**
 * DISTRIBUTED DATABASE ARCHITECTURE - JAVA IMPLEMENTATION
 */

public class DistributedDatabase {
    
    enum NodeRole { LEADER, FOLLOWER }
    
    static class ReplicaNode {
        int nodeId;
        String address;
        NodeRole role;
        int term;
        int logIndex;
        int commitmentsAcked;
        
        ReplicaNode(int id, String addr) {
            this.nodeId = id;
            this.address = addr;
            this.role = NodeRole.FOLLOWER;
            this.term = 0;
            this.logIndex = 0;
            this.commitmentsAcked = 0;
        }
    }
    
    static class LogEntry {
        int entryId;
        int term;
        String data;
        
        LogEntry(int id, int t, String d) {
            this.entryId = id;
            this.term = t;
            this.data = d;
        }
    }
    
    static class DDB {
        List<ReplicaNode> replicas;
        List<LogEntry> log;
        int totalWrites;
        int totalReads;
        int quorumSize;
        
        DDB(int replicaCount) {
            this.replicas = new ArrayList<>();
            this.log = new ArrayList<>();
            this.totalWrites = 0;
            this.totalReads = 0;
            this.quorumSize = (replicaCount / 2) + 1;
            
            for (int i = 0; i < replicaCount; i++) {
                ReplicaNode node = new ReplicaNode(i + 1, "db" + (i + 1) + ".local");
                if (i == 0) node.role = NodeRole.LEADER;
                replicas.add(node);
            }
            
            System.out.println("Distributed Database created with " + replicaCount + 
                             " replicas (quorum: " + quorumSize + ")");
        }
        
        void write(String data) {
            System.out.println("\n[Write] " + data);
            
            log.add(new LogEntry(log.size() + 1, replicas.get(0).term, data));
            totalWrites++;
            
            int acked = 0;
            for (ReplicaNode replica : replicas) {
                System.out.println("  → Replicating to Replica " + replica.nodeId + 
                                 ": " + data);
                replica.commitmentsAcked++;
                acked++;
            }
            
            System.out.println("  Replicated to " + acked + " replicas (quorum: " + 
                             quorumSize + " achieved)");
        }
        
        void read(int entryId) {
            System.out.println("\n[Read] Entry " + entryId);
            totalReads++;
            
            if (entryId > log.size()) {
                System.out.println("  Entry not found");
                return;
            }
            
            LogEntry entry = log.get(entryId - 1);
            System.out.println("  Data: " + entry.data + " (term: " + entry.term + ")");
        }
        
        void printStats() {
            System.out.println("\n╔════════════════════════════════════════════════════════════════╗");
            System.out.println("║ DISTRIBUTED DATABASE STATISTICS                              ║");
            System.out.println("╚════════════════════════════════════════════════════════════════╝");
            
            System.out.println("\nDatabase Status:");
            System.out.println("  Total replicas: " + replicas.size());
            System.out.println("  Quorum size: " + quorumSize);
            System.out.println("  Total writes: " + totalWrites);
            System.out.println("  Total reads: " + totalReads);
            System.out.println("  Log size: " + log.size());
            
            System.out.println("\nReplica Status:");
            System.out.println(String.format("  %-10s %-20s %-10s %-10s",
                                           "ID", "Address", "Role", "Acked"));
            System.out.println(String.format("  %-10s %-20s %-10s %-10s",
                                           "---", "-------", "----", "-----"));
            
            for (ReplicaNode r : replicas) {
                System.out.println(String.format("  %-10d %-20s %-10s %-10d",
                                                r.nodeId, r.address,
                                                r.role, r.commitmentsAcked));
            }
        }
    }
    
    static void demo() {
        System.out.println("╔════════════════════════════════════════════════════════════════╗");
        System.out.println("║ DISTRIBUTED DATABASE - COMPREHENSIVE DEMONSTRATION           ║");
        System.out.println("╚════════════════════════════════════════════════════════════════╝");
        
        DDB ddb = new DDB(3);
        
        System.out.println("\n1. Perform Writes (Replicated):");
        System.out.println("──────────────────────────────");
        
        ddb.write("INSERT users VALUES ('Alice', 25)");
        ddb.write("INSERT users VALUES ('Bob', 30)");
        ddb.write("UPDATE users SET age=26 WHERE name='Alice'");
        
        System.out.println("\n2. Perform Reads:");
        System.out.println("─────────────────");
        
        ddb.read(1);
        ddb.read(2);
        ddb.read(3);
        
        ddb.printStats();
    }
    
    public static void main(String[] args) {
        demo();
        
        System.out.println("\n╔════════════════════════════════════════════════════════════════╗");
        System.out.println("║ KEY PATTERNS                                                 ║");
        System.out.println("╚════════════════════════════════════════════════════════════════╝");
        
        System.out.println("\n1. Raft Consensus: Leader election");
        System.out.println("2. Quorum Reads/Writes: Ensure consistency");
        System.out.println("3. Two-Phase Commit: Distributed transactions");
        System.out.println("4. Log Replication: Durability");
        System.out.println("5. Failover: Automatic recovery");
    }
}
